### PR TITLE
Feature/44 HTTP interceptor

### LINKED ISSUE #44: HTTP interсeptor
Для начала нужно ознакомиться что такое интерсептор. Материалы к прочтению:
- https://dev.to/teroauralinna/global-http-request-and-response-handling-with-the-axios-interceptor-30ae
- https://medium.com/swlh/handling-access-and-refresh-tokens-using-axios-interceptors-3970b601a5da
- https://medium.com/@yaob/how-to-globally-use-axios-instance-and-interceptors-e28f351bb794

В рамках задачи, нужно сделать:

1. `ApiService` (`api-service.service.js`), который выполняет:
- [x] работа с заголовками запросов
- [x] подставляет базовый урл апи из настроек в адресс для запроса
- [x] предоставляет простой интерфейс для запросов на api, через который будут проходить все запросы

```
$api.get('/path', data)
$api.post('/path', data)
$api.put('/path', data)
$api.delete('/path', data)
```
где, например, вместо `'/path'` может стоять `'/userSettings'`, а `data` - просто объект с данными для запросами `{userId: 2, firstName: 'Anastasiia', .. }`

2. `ApiHttpInterceptor` - интерсептор для апи сервиса (`api-http-interceptor.service.js`), который:
- [x] берет на себя работу с токенами
- [x] обрабатывает коды ошибок

3. Удалить из всего остального кода:
- [x] все прямые вызовы апи  и заменить на вызовы ApiService через `$api`
- [x] управление токенами
- [x] добавление заголовков
- [x] обработку ошибок

4. Конечно, проверить что ничего не поломалось. Если поломалось - починить

можно добавить директорию `@services`

### PR DESCRIPTION
**Pull request by ivhorodko**
#44

### CODE REVIEW

File: front/src/services/api-http-interceptor.service.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        baseURL: baseURL,
+        headers: {
+          Accept: 'application/json',
+          'Content-Type': 'application/json'
+        }
+      });
+
+instance.interceptors.request.use(
+  config => {
+    const token = JSON.parse(localStorage.getItem('user'));
```
Comment (sanioooook) [inline front/src/services/api-http-interceptor.service.js:15]: why do you take this value directly and not from the store?

Comment (ivhorodko) [reply front/src/services/api-http-interceptor.service.js:15]: https://github.com/horba/my-trip/issues/44
point 2

File: front/src/store/modules/trip.js [hunk 2]
```diff
@@ -18,15 +15,15 @@ export default {
...
           if (r.status === 200) {
             commit('INIT_TRIPS_HISTORY', r.data);
           }
         });
     },
     async searchTripsHistory ({ commit }, searchQuery) {
-      await axios.get(`${serverPath}/api/trips/history?searchQuery=${searchQuery}`)
+      await api.get(`/trips/history?searchQuery=${searchQuery}`)
         .then(r => {
           if (r.status === 200) {
```
Comment (horba) [inline front/src/store/modules/trip.js]: it is unnecessary check

File: front/src/store/index.js [hunk 3]
```diff
@@ -23,17 +20,15 @@ export default new Vuex.Store({
...
     CLEAR_USER_DATA (state) {
       localStorage.removeItem('user');
-      delete axios.defaults.headers.common.Authorization;
       location.reload();
     }
   },
   actions: {
     async signUp (context, body) {
-      return await axios.post(`${serverPath}/api/auth/signup`, body)
+      return await api.post('/auth/signup', body)
```
Comment (horba) [inline front/src/store/index.js:33]: doesn't it work this.$api.post() etc?

File: front/src/config/config.dev.json [hunk 4]
```diff
@@ -1,3 +1,4 @@
 {
-  "serverPath": "https://localhost:5001"
+  "serverPath": "https://localhost:5001",
+  "apiPath": "https://localhost:5001/api"
```
Comment (horba) [inline front/src/config/config.dev.json]: usually it is stored serverPath(baseUrl is usual name) and api suffix

```
  "baseUrl": "https://localhost:5001",
  "apiPath": "api" //or "apiSuffix": "api"
```

File: front/src/services/api-http-interceptor.service.js (hunk 1 above)
Comment (sanioooook) [reply front/src/services/api-http-interceptor.service.js:15]: got it

**Review by horba — CHANGES_REQUESTED** [review]
also it should be checked with new dev - it was added new functionality of user avatar and it works with images and files and form data - it is really important to redo
decide if you want use globally $api - use it, don`t import , if no - remove $api from vue

File: front/src/config/config.dev.json [hunk 5]
```diff
@@ -1,3 +1,4 @@
 {
-  "serverPath": "https://localhost:5001"
+  "baseUrl": "https://localhost:5001",
+  "apiSuffix": "/api"
```
Comment (horba) [inline front/src/config/config.dev.json]: nit - just "api" no backslash

expressing that something could be better, but if it’s not very important, prefix it with something like “Nit: “ to let the author know that it’s just a point of polish that they could choose to ignore. source: https://google.github.io/eng-practices/review/reviewer/standard.html

File: front/src/services/api-http-interceptor.service.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  response => {
+    if (response.status === 200 || response.status === 201) {
+      return Promise.resolve(response);
+    } else {
+      return Promise.reject(response);
+    }
+  }, error => {
+    if (error.response.status) {
+      switch (error.response.status) {
+        case 400:
```
Comment (demosglok) [inline front/src/services/api-http-interceptor.service.js]: why do we need this "case" if it acts exactly the same as default return

Comment (ivhorodko) [reply front/src/services/api-http-interceptor.service.js]: This is an example for the guys to understand what is happening here. But I can remove it

Comment (ivhorodko) [reply front/src/services/api-http-interceptor.service.js]: It just seemed to me that it would be clearer

Comment (horba) [reply front/src/services/api-http-interceptor.service.js]: yes, it seems it could be removed and merged))) I dream to merge it

**Review by horba — APPROVED** [review]

**Pull request merged**